src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/CachedBinarySearchIntervalList.java

```diff
@@ -0,0 +1,161 @@
+package org.broadinstitute.hellbender.tools.copynumber.utils;
+
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.tools.exome.HashedListTargetCollection;
+import org.broadinstitute.hellbender.utils.IndexRange;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Collection of {@link Locatable}s that supports cached binary search of {@link Locatable} objects. In particular,
+ * this class provides an interface to get a range of {@link Locatable}s that intersects with a given {@link Locatable}
+ * query location.
+ *
+ * The search is optimized based on the assumption that the incoming queries of {@link Locatable}s are in sorted order.
+ *
+ * @param <E> a locatable object
+ */
+public class CachedBinarySearchIntervalList<E extends Locatable> {
```

Comment:
As we discussed, I would be curious what the actual performance hit would be if you simply used an OverlapDetector from htsjdk.